### ISSUE TITLE
[Bug]: Channel cp lag a lot after some pods terminated due to disconnected from etcd and restarted

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#30823]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: cardinal-milvus-io-2.3-ef086dc-20240222
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar   
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. milvus-cluster has 2 dataNode, 5 indexNode, 4 queryNode
2. collection laion_stable_4 has about 60m-768d data and it's schema is:
```
{'auto_id': False, 'description': '', 'fields': [{'name': 'id', 'description': '', 'type': <DataType.INT64: 5>, 'is_primary': True, 'auto_id': False}, {'name': 'float_vector', 'description': '', 'type': <DataType.FLOAT_VECTOR: 101>, 'params': {'dim': 768}}, {'name': 'int64_pk_5b', 'description': '', 'type': <DataType.INT64: 5>, 'is_partition_key': True}, {'name': 'varchar_caption', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 8192}}, {'name': 'varchar_NSFW', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 8192}}, {'name': 'float64_similarity', 'description': '', 'type': <DataType.FLOAT: 10>}, {'name': 'int64_width', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'int64_height', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'int64_original_width', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'int64_original_height', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'varchar_md5', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 8192}}], 'enable_dynamic_field': True}
```
3. concurrent requests: insert + delete + flush + search
4. 1 dataNode, 4 indexNode and 1 queryNode terminated due to disconnect from etcd
[qn_jpf7x_pre.log](https://github.com/milvus-io/milvus/files/14401539/qn_jpf7x_pre.log)
[in_lmnnc_pre.log](https://github.com/milvus-io/milvus/files/14401540/in_lmnnc_pre.log)
[dn_6c5sl_pre.log](https://github.com/milvus-io/milvus/files/14401543/dn_6c5sl_pre.log)

```
[2024/02/24 14:43:50.946 +00:00] [INFO] [datanode/flush_manager.go:966] ["successfully save binlog"] [segmentID=447902879719100167] [collectionID=447902879639453431] [vchannel=laion1b-test-2-rootcoord-dml_4_447902879639453431v0]
[2024/02/24 14:43:51.116 +00:00] [WARN] [sessionutil/session_util.go:548] ["fail to retry keepAliveOnce"] [serverName=datanode] [LeaseID=102048947326302981] [error="attempt #0: context deadline exceeded: attempt #1: etcdserver: requested lease not found: attempt #2: etcdserver: requested lease not found"]
[2024/02/24 14:43:51.117 +00:00] [WARN] [sessionutil/session_util.go:876] ["connection lost detected, shuting down"]
[2024/02/24 14:43:51.117 +00:00] [ERROR] [datanode/data_node.go:189] ["Data Node disconnected from etcd, process will exit"] ["Server Id"=1629] [stack="github.com/milvus-io/milvus/internal/datanode.(*DataNode).Register.func1\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/data_node.go:189"]
```

dataCoord cp lag:
![image](https://github.com/milvus-io/milvus/assets/27288593/d1e147c4-78d2-4f27-9c3a-bf5e2033f670)

### Milvus Log

- argo
1. https://argo-workflows.zilliz.cc/archived-workflows/qa/e24ade9c-aab7-4958-ac79-7ba8e668d10f?nodeId=laion1b-test-cron-1708747260
2. https://argo-workflows.zilliz.cc/archived-workflows/qa/c5443b7d-5a1a-45fd-81b3-4881fda1933f?nodeId=laion1b-test-cron-1708790460
3. https://argo-workflows.zilliz.cc/archived-workflows/qa/e9cde1ad-54bc-443c-ab67-af60da10b8c4?nodeId=laion1b-test-cron-1708833660
4. https://argo-workflows.zilliz.cc/archived-workflows/qa/902d6640-6339-4c16-895b-7287c9d37b10

- pods:
```
laion1b-test-2-etcd-0                                             1/1     Running       1 (45d ago)     73d     10.104.25.31    4am-node30   <none>           <none>
laion1b-test-2-etcd-1                                             1/1     Running       0               73d     10.104.30.94    4am-node38   <none>           <none>
laion1b-test-2-etcd-2                                             1/1     Running       0               73d     10.104.34.225   4am-node37   <none>           <none>
laion1b-test-2-milvus-datanode-648fd8c79c-6c5sl                   1/1     Running       1 (41h ago)     3d17h   10.104.9.228    4am-node14   <none>           <none>
laion1b-test-2-milvus-datanode-648fd8c79c-psk8h                   1/1     Running       0               3d17h   10.104.15.115   4am-node20   <none>           <none>
laion1b-test-2-milvus-indexnode-fffb75cd7-8c7mt                   1/1     Running       2 (41h ago)     3d16h   10.104.17.173   4am-node23   <none>           <none>
laion1b-test-2-milvus-indexnode-fffb75cd7-9ssh7                   1/1     Running       0               3d16h   10.104.26.104   4am-node32   <none>           <none>
laion1b-test-2-milvus-indexnode-fffb75cd7-lmnnc                   1/1     Running       1 (41h ago)     3d16h   10.104.5.134    4am-node12   <none>           <none>
laion1b-test-2-milvus-indexnode-fffb75cd7-pb6nt                   1/1     Running       1 (41h ago)     3d16h   10.104.9.230    4am-node14   <none>           <none>
laion1b-test-2-milvus-indexnode-fffb75cd7-x8gdm                   1/1     Running       1 (41h ago)     3d16h   10.104.29.169   4am-node35   <none>           <none>
laion1b-test-2-milvus-mixcoord-676bd65bcf-vqlkf                   1/1     Running       0               3d17h   10.104.18.76    4am-node25   <none>           <none>
laion1b-test-2-milvus-proxy-db5b89665-j9xsr                       1/1     Running       0               3d17h   10.104.27.159   4am-node31   <none>           <none>
laion1b-test-2-milvus-querynode-1-5956458d77-4bntq                1/1     Running       0               3d16h   10.104.32.44    4am-node39   <none>           <none>
laion1b-test-2-milvus-querynode-1-5956458d77-bt7dt                1/1     Running       0               3d16h   10.104.28.226   4am-node33   <none>           <none>
laion1b-test-2-milvus-querynode-1-5956458d77-jpf7x                1/1     Running       1 (41h ago)     3d16h   10.104.17.175   4am-node23   <none>           <none>
laion1b-test-2-milvus-querynode-1-5956458d77-qw8s6                1/1     Running       0               3d16h   10.104.33.242   4am-node36   <none>           <none>
laion1b-test-2-pulsar-bookie-0                                    1/1     Running       0               73d     10.104.33.107   4am-node36   <none>           <none>
laion1b-test-2-pulsar-bookie-1                                    1/1     Running       0               49d     10.104.18.240   4am-node25   <none>           <none>
laion1b-test-2-pulsar-bookie-2                                    1/1     Running       0               73d     10.104.25.32    4am-node30   <none>           <none>
laion1b-test-2-pulsar-broker-0                                    1/1     Running       0               67d     10.104.1.69     4am-node10   <none>           <none>
laion1b-test-2-pulsar-proxy-0                                     1/1     Running       0               73d     10.104.4.218    4am-node11   <none>           <none>
laion1b-test-2-pulsar-recovery-0                                  1/1     Running       0               73d     10.104.14.151   4am-node18   <none>           <none>
laion1b-test-2-pulsar-zookeeper-0                                 1/1     Running       0               73d     10.104.29.87    4am-node35   <none>           <none>
laion1b-test-2-pulsar-zookeeper-1                                 1/1     Running       0               73d     10.104.21.124   4am-node24   <none>           <none>
laion1b-test-2-pulsar-zookeeper-2                                 1/1     Running       0               73d     10.104.34.229   4am-node37   <none>           <none>
```

- grafana: [metrics of laion1b-test-2](https://grafana-4am.zilliz.cc/d/uLf5cJ3Ga/milvus2-0?orgId=1&from=1708771953238&to=1708934074587&var-datasource=P1809F7CD0C75ACF3&var-cluster=&var-namespace=qa-milvus&var-instance=laion1b-test-2&var-collection=All&var-app_name=milvus)

**Comment by ThreadDao** [comment]
Is it expected that I haven’t done any compaction during this period?
![image](https://github.com/milvus-io/milvus/assets/27288593/83052b9f-91af-4894-a250-9a9df1f0c2c3)

**Comment by chyezh** [comment]
Found that new datanode reloaded cannot watch channel.

```
[2024/02/24 15:33:52.121 +00:00] [WARN] [datanode/flow_graph_manager.go:136] ["fail to create new DataSyncService"] [channel=laion1b-test-2-rootcoord-dml_5_447902879639453431v1] [error="tickler watch failed"] [errorVerbose="tickler watch failed\n(1) attached stack trace\n  -- stack trace:\n  | github.com/milvus-io/milvus/internal/datanode.getChannelWithEtcdTickler\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/data_sync_service.go:291\n  | github.com/milvus-io/milvus/internal/datanode.newServiceWithEtcdTickler\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/data_sync_service.go:426\n  | github.com/milvus-io/milvus/internal/datanode.(*flowgraphManager).addAndStartWithEtcdTickler\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_manager.go:131\n  | github.com/milvus-io/milvus/internal/datanode.(*DataNode).handlePutEvent\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:179\n  | github.com/milvus-io/milvus/internal/datanode.(*channelEventManager).Run.func1\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:268\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1598\nWraps: (2) tickler watch failed\nError types: (1) *withstack.withStack (2) *errutil.leafError"]

[2024/02/24 15:33:54.035 +00:00] [WARN] [datanode/flow_graph_manager.go:136] ["fail to create new DataSyncService"] [channel=laion1b-test-2-rootcoord-dml_4_447902879639453431v0] [error="tickler watch failed"] [errorVerbose="tickler watch failed\n(1) attached stack trace\n  -- stack trace:\n  | github.com/milvus-io/milvus/internal/datanode.getChannelWithEtcdTickler\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/data_sync_service.go:291\n  | github.com/milvus-io/milvus/internal/datanode.newServiceWithEtcdTickler\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/data_sync_service.go:426\n  | github.com/milvus-io/milvus/internal/datanode.(*flowgraphManager).addAndStartWithEtcdTickler\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_manager.go:131\n  | github.com/milvus-io/milvus/internal/datanode.(*DataNode).handlePutEvent\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:179\n  | github.com/milvus-io/milvus/internal/datanode.(*channelEventManager).Run.func1\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:268\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1598\nWraps: (2) tickler watch failed\nError types: (1) *withstack.withStack (2) *errutil.leafError"]

```

**Comment by chyezh** [comment]
Found that datanode always recover channel failure for timeout.

```
2024-02-24 22:53:40.149	
(no unique labels)
[2024/02/24 14:53:40.149 +00:00] [INFO] [datanode/event_manager.go:128] ["DataNode is handling watchInfo PUT event"] [key=laion1b-test-2/meta/channelwatch/1660/laion1b-test-2-rootcoord-dml_4_447902879639453431v0] ["watch state"=ToWatch]
2024-02-24 22:58:40.953	
(no unique labels)
[2024/02/24 14:58:40.953 +00:00] [INFO] [datanode/event_manager.go:128] ["DataNode is handling watchInfo PUT event"] [key=laion1b-test-2/meta/channelwatch/1660/laion1b-test-2-rootcoord-dml_4_447902879639453431v0] ["watch state"=ToRelease]
2024-02-24 23:01:33.745	
(no unique labels)
[2024/02/24 15:01:33.744 +00:00] [ERROR] [datanode/event_manager.go:356] ["tickler update failed: failed to compare version and swap"] [key=channelwatch/1660/laion1b-test-2-rootcoord-dml_4_447902879639453431v0] [progress=281] [version=1] [vChanName=laion1b-test-2-rootcoord-dml_4_447902879639453431v0] [stack="github.com/milvus-io/milvus/internal/datanode.(*etcdTickler).watch.func1\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:356"]
2024-02-24 23:03:55.513	
(no unique labels)
[2024/02/24 15:03:55.513 +00:00] [INFO] [datanode/event_manager.go:131] ["DataNode is handling watchInfo DELETE event"] [key=laion1b-test-2/meta/channelwatch/1660/laion1b-test-2-rootcoord-dml_4_447902879639453431v0]
2024-02-24 23:13:47.290	
(no unique labels)
[2024/02/24 15:13:47.290 +00:00] [INFO] [datanode/event_manager.go:128] ["DataNode is handling watchInfo PUT event"] [key=laion1b-test-2/meta/channelwatch/1660/laion1b-test-2-rootcoord-dml_4_447902879639453431v0] ["watch state"=ToWatch]
2024-02-24 23:18:48.729	
(no unique labels)
[2024/02/24 15:18:48.729 +00:00] [INFO] [datanode/event_manager.go:128] ["DataNode is handling watchInfo PUT event"] [key=laion1b-test-2/meta/channelwatch/1660/laion1b-test-2-rootcoord-dml_4_447902879639453431v0] ["watch state"=ToRelease]
```
Too much sealed segments to recover from cp.

```
[2024/02/24 15:05:40.074 +00:00] [INFO] [datanode/data_sync_service.go:260] ["recover sealed segments form checkpoints"] [vChannelName=laion1b-test-2-rootcoord-dml_4_447902879639453431v0] [segmentID=447902879717839140] [numRows=65]
```

@ThreadDao Please modify the watch channel timeout to 30 min recover channel.

We need use rpc-based channel watch to handle it.
Channel watch timeout should based on a heartbeat between  datanode and datacoord.
But not a status change.

**Comment by chyezh** [comment]
Serious lock contention on meta of datacoords.
An enhancement should be applied to avoid this issue.

**Comment by chyezh** [comment]
Found some Dropping collections cannot be dropped.

- Dropped collection cannot be recovered at datanode.
- segment on it cannot be marked as droped by datanode (446771525888510903
447180040934991941
447619508238296599).

GC is very slow, like #29655

**Comment by chyezh** [comment]
Using typeutil.Set leads to poor performance in massive retrieval of `GetSegmentInfo`.

<img width="970" alt="image" src="https://github.com/milvus-io/milvus/assets/20332315/2ba1d31d-09d8-4500-9b8c-9d533780629c">

**Comment by chyezh** [comment]
There'are many codes with pattern `for .* segments` can be found in datacoord/meta.go.
Should be optimized if the function is applied in a for-loop.

**Comment by xiaofan-luan** [comment]
maybe it's reasonable to create some inverted index on top of it

**Comment by ThreadDao** [comment]
fixed `milvus-io-2.3-fef430d-20240320`
